server/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java

```diff
@@ -375,42 +375,61 @@ public void messageReceived(final NodeRequest request, TransportChannel channel)
             if (logger.isTraceEnabled()) {
                 logger.trace("[{}] executing operation on [{}] shards", actionName, totalShards);
             }
-            final Object[] shardResultOrExceptions = new Object[totalShards];
+            List<BroadcastShardOperationFailedException> accumulatedExceptions = Collections.synchronizedList(new ArrayList<>());
+            List<ShardOperationResult> results = Collections.synchronizedList(new ArrayList<>());
 
-            int shardIndex = -1;
-            for (final ShardRouting shardRouting : shards) {
-                shardIndex++;
-                onShardOperation(request, shardResultOrExceptions, shardIndex, shardRouting);
-            }
+            ActionListener<ShardOperationResult> listener = new ActionListener<ShardOperationResult>() {
```

Comment:
Note that we can't use MultiActionListener here because we want to collect all shard failures and return them separately, rather than failing the whole request.